pyro/nn/module.py

```diff
@@ -0,0 +1,196 @@
+PyroSample.__doc__ = """
+Structure to declare a Pyro-managed random parameter of a :class:`PyroModule`.
+"""
+
+
+def _make_name(prefix, name):
+    return "{}.{}".format(prefix, name) if prefix else name
+
+
+class PyroModule(torch.nn.Module):
+    """
+    Subclass of :class:`torch.nn.Module` that supports setting of
+    :class:`PyroParam` and :class:`PyroSample` .
+
+    To create a Pyro-managed parameter attribute, set that attribute using the
+    :class:`PyroParam` helper. Reading that attribute will then trigger a
+    :func:`pyro.param` statement. For example::
+
+        # Create Pyro-managed parameter attributes.
+        my_module = PyroModule()
+        my_module.loc = PyroParam(torch.tensor(0.))
+        my_module.scale = PyroParam(torch.tensor(1.),
+                                    constraint=constraints.positive)
+        # Read the attributes.
+        loc = my_module.loc  # Triggers a pyro.param statement.
+        scale = my_module.scale  # Triggers another pyro.param statement.
+
+    Note that, unlike normal :class:`torch.nn.Module` s, :class:`PyroModule` s
+    should note be registered with :func:`pyro.module` statements.
+    :class:`PyroModule` s can contain normal :class:`torch.nn.Module` s, but not
+    vice versa. Accessing a normal :class:`torch.nn.Module` attribute of
+    a :class:`PyroModule` triggers a :func:`pyro.module` statement.
+
+    To create a Pyro-managed random attribute, set that attribute using the
+    :class:`PyroSample` helper, specifying a prior distribution. Reading that
+    attribute will then trigger a :func:`pyro.sample` statement. For example::
+
+        # Create Pyro-managed random attributes.
+        my_module.x = PyroSample(dist.Normal(0, 1))
+        my_module.y = PyroSample(lambda self: dist.Normal(self.loc, self.scale))
+
+        # Sample the attributes.
+        x = my_module.x  # Triggers a pyro.sample statement.
+        y = my_module.y  # Triggers one pyro.sample + two pyro.param statements.
+
+    Note that, because sample statements can appear only once in a trace, you
+    can read the value of a random attribute only once per inference step.
+
+    To make an existing module probabilistic, you can create a subclass and
+    overwrite some parameters with :class:`PyroSample` s::
+
+        class RandomLinear(nn.Linear, PyroModule):
+            def __init__(in_features, out_features):
+                super().__init__(self, in_features, out_features)
+                self.weight = PyroSample(
+                    lambda self: dist.Normal(0, 1)
+                                     .expand([self.out_features,
+                                              self.in_features])
+                                     .to_event(2))
+    """
+    def __init__(self):
+        self._pyro_name = ""
+        self._pyro_params = OrderedDict()
+        self._pyro_samples = OrderedDict()
+        super().__init__()
+
+    def _pyro_set_name(self, name):
+        self._pyro_name = name
```

Comment:
While not really necessary, it will be nice if the pyro names corresponded to the module / param names that we get from `module.named_parameters` or `module.named_modules`.

Reply:
This should be the case, but I'll add a test.

Reply:
done.